### PR TITLE
[python-package] fix mypy errors about custom eval and metric functions

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #3867.

Fixes the following `mypy` errors:

```text
engine.py:258: error: Argument 1 to "eval_train" of "Booster" has incompatible type "Union[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]], List[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]]], None]"; expected "Union[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]], List[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]]], None]"  [arg-type]
engine.py:259: error: Argument 1 to "eval_valid" of "Booster" has incompatible type "Union[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]], List[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]]], None]"; expected "Union[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]], List[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]]], None]"  [arg-type]

sklearn.py:130: error: Too few arguments  [call-arg]
sklearn.py:130: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:132: error: Too many arguments  [call-arg]
sklearn.py:132: error: Too few arguments  [call-arg]
sklearn.py:132: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:132: error: Argument 3 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:134: error: Too many arguments  [call-arg]
sklearn.py:134: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:134: error: Argument 3 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:134: error: Argument 4 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:208: error: Too few arguments  [call-arg]
sklearn.py:208: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:210: error: Too many arguments  [call-arg]
sklearn.py:210: error: Too few arguments  [call-arg]
sklearn.py:210: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:210: error: Argument 3 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:212: error: Too many arguments  [call-arg]
sklearn.py:212: error: Argument 1 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:212: error: Argument 3 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:212: error: Argument 4 has incompatible type "Optional[ndarray[Any, Any]]"; expected "ndarray[Any, Any]"  [arg-type]
sklearn.py:814: error: Argument "feval" to "train" has incompatible type "List[_EvalFunctionWrapper]"; expected "Union[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]], List[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]]], None]"  [arg-type]
sklearn.py:814: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
sklearn.py:814: note: Consider using "Sequence" instead, which is covariant
```

### CODE REVIEW

File: python-package/lightgbm/sklearn.py [hunk 1]
```diff
@@ -25,32 +25,50 @@
 ]
 
 _LGBM_ScikitCustomObjectiveFunction = Union[
+    # f(labels, preds)
     Callable[
-        [np.ndarray, np.ndarray],
+        [Optional[np.ndarray], np.ndarray],
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:30]: ### why add all these `Optional`s?

`mypy` is struggling with the facts that `Dataset.get_field()` can return `None`.
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/basic.py#L2704
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/sklearn.py#L127
* https://github.com/microsoft/LightGBM/blob/2fe2bf0675ef127a29f9bc49511fc0099cf3e140/python-package/lightgbm/basic.py#L2393

This PR proposes updating the type hints for custom metric and objective functions to match that behavior.

I intentionally chose not to update the user-facing docs about custom metric and objective functions to reflect that the `label`, `group`, and `weights` passed to these functions can technically be `None`... in almost all situations, they should be non-`None`. I don't think complicating the docs is worth it.

File: python-package/lightgbm/sklearn.py [hunk 2]
```diff
@@ -25,32 +25,50 @@
...
+    ],
+    # f(labels, preds, weights, group)
+    Callable[
+        [Optional[np.ndarray], np.ndarray, Optional[np.ndarray], Optional[np.ndarray]],
+        _LGBM_EvalFunctionResultType
+    ],
+    Callable[
+        [Optional[np.ndarray], np.ndarray, Optional[np.ndarray], Optional[np.ndarray]],
+        List[_LGBM_EvalFunctionResultType]
+    ]
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:71]: Splitting cases like

```python
Callable[
    [np.ndarray, np.ndarray],
    Union[_LGBM_EvalFunctionResultType, List[_LGBM_EvalFunctionResultType]]
]
```

into

```python
Union[
    Callable[
        [np.ndarray, np.ndarray],
        _LGBM_EvalFunctionResultType
    ],
    Callable[
        [np.ndarray, np.ndarray],
        List[_LGBM_EvalFunctionResultType]
    ]
]
```

helps `mypy` with errors like this:

```text
Argument 1 to "eval_valid" of "Booster" has incompatible type "Union[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]], List[Callable[[ndarray[Any, Any], Dataset], Union[Tuple[str, float, bool], List[Tuple[str, float, bool]]]]], None]"; expected "Union[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]], List[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]]], None]"  [arg-type]
```

And I think it's slightly more correct... I'd expect people to provide a custom metric function that returns a single tuple on every iteration or one that returns a list of tuples on each iteration, but not one that could return *either of those*.

File: python-package/lightgbm/sklearn.py [hunk 3]
```diff
@@ -127,11 +145,11 @@ def __call__(self, preds: np.ndarray, dataset: Dataset) -> Tuple[np.ndarray, np.
...
         argc = len(signature(self.func).parameters)
         if argc == 2:
-            grad, hess = self.func(labels, preds)
+            grad, hess = self.func(labels, preds)  # type: ignore[call-arg]
         elif argc == 3:
-            grad, hess = self.func(labels, preds, dataset.get_weight())
+            grad, hess = self.func(labels, preds, dataset.get_weight())  # type: ignore[call-arg]
         elif argc == 4:
-            grad, hess = self.func(labels, preds, dataset.get_weight(), dataset.get_group())
+            grad, hess = self.func(labels, preds, dataset.get_weight(), dataset.get_group())  # type: ignore [call-arg]
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:152]: `mypy` isn't able to break apart the individual `Union` items base on how many arguments `inspect.signature()` returns. This PR proposes just ignoring errors like this:

```text
sklearn.py:132: error: Too many arguments  [call-arg]
sklearn.py:132: error: Too few arguments  [call-arg]
```

File: python-package/lightgbm/sklearn.py [hunk 4]
```diff
@@ -811,7 +829,7 @@ def _get_meta_data(collection, name, i):
             num_boost_round=self.n_estimators,
             valid_sets=valid_sets,
             valid_names=eval_names,
-            feval=eval_metrics_callable,
+            feval=eval_metrics_callable,  # type: ignore[arg-type]
```
Comment (jameslamb) [inline python-package/lightgbm/sklearn.py:832]: This `# type: ignore` fixes this error:

```text
Argument "feval" to "train" has incompatible type "List[_EvalFunctionWrapper]"; expected "Union[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]], List[Union[Callable[[ndarray[Any, Any], Dataset], Tuple[str, float, bool]], Callable[[ndarray[Any, Any], Dataset], List[Tuple[str, float, bool]]]]], None]"  [arg-type]
sklearn.py:814: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
sklearn.py:814: note: Consider using "Sequence" instead, which is covariant
```

I'm not sure why, but `mypy` isn't able to figure out that `_EvalFunctionWrapper` is actually a `Callable[[ndarray, Dataset], Tuple[str, float, bool]]`.

I think that's maybe related to some issues it has with comparing lists to union types including lists? e.g. https://github.com/python/mypy/issues/6463

This PR proposes just ignoring that error for now... the use of custom eval metric functions is well-covered by the unit tests in `test_sklearn.py` and `test_dask.py`.

**Comment by jameslamb** [comment]
I was looking into some other `mypy` errors tonight and they required touching the same type annotations as this PR I'd opened a few days ago. So I just pushed those changes here, in https://github.com/microsoft/LightGBM/pull/5790/commits/d0d8d9ebf33c25736f888701333dea2cd2b6e709.

This is ready for review.

**Review by guolinke — APPROVED** [review]
